internal/timecmp/cmp_test.go

```diff
@@ -0,0 +1,269 @@
+		})
+
+		// x after y
+		runTest(t, tc{
+			x:           b,
+			y:           a,
+			cmp:         BeforeOrEqual,
+			nanosecond:  false,
+			microsecond: false,
+			second:      false,
+		})
+	})
+}
+
+func TestAfterOrEqual(t *testing.T) {
+	reflexiveTest(t, AfterOrEqual)
+
+	t.Run("NanosecondDifference", func(t *testing.T) {
+		a := time.Now().Truncate(time.Nanosecond).Add(-5 * time.Nanosecond)
+		b := a.Add(time.Nanosecond)
+
+		// x before y
+		runTest(t, tc{
+			x:           a,
+			y:           b,
+			cmp:         AfterOrEqual,
+			nanosecond:  false,
+			microsecond: true,
+			second:      true,
+		})
+
+		// x after y
+		runTest(t, tc{
+			x:           b,
+			y:           a,
+			cmp:         AfterOrEqual,
+			nanosecond:  true,
+			microsecond: true,
+			second:      true,
+		})
+	})
+
+	t.Run("MicrosecondDifference", func(t *testing.T) {
+		a := time.Now().Truncate(time.Microsecond).Add(-5 * time.Microsecond)
+		b := a.Add(time.Microsecond)
+
+		// x before y
+		runTest(t, tc{
+			x:           a,
+			y:           b,
+			cmp:         AfterOrEqual,
+			nanosecond:  false,
+			microsecond: false,
+			second:      true,
+		})
+
+		// x after y
+		runTest(t, tc{
+			x:           b,
+			y:           a,
+			cmp:         AfterOrEqual,
+			nanosecond:  true,
+			microsecond: true,
+			second:      true,
+		})
+	})
+
+	t.Run("SecondDifference", func(t *testing.T) {
+		a := time.Now().Truncate(time.Second).Add(-5 * time.Second)
+		b := a.Add(time.Second)
+
+		// x before y
+		runTest(t, tc{
+			x:           a,
+			y:           b,
+			cmp:         AfterOrEqual,
+			nanosecond:  false,
+			microsecond: false,
+			second:      false,
+		})
+
+		// x after y
+		runTest(t, tc{
+			x:           b,
+			y:           a,
+			cmp:         AfterOrEqual,
+			nanosecond:  true,
+			microsecond: true,
+			second:      true,
+		})
+	})
+}
+
+func reflexiveTest(t *testing.T, cmp cmpFunc) {
+	t.Run("Reflexive", func(t *testing.T) {
+		now := time.Now()
+		runTest(t, tc{
+			x:           now,
+			y:           now,
+			cmp:         cmp,
+			nanosecond:  true,
+			microsecond: true,
+			second:      true,
+		})
+	})
+}
+
+func runTest(t testing.TB, tc tc) {
+	assert.Equal(t, tc.nanosecond, tc.cmp(tc.x, tc.y), "Nanosecond (stdlib <> stdlib) comparison failed")
+
+	assert.Equal(t, tc.microsecond, tc.cmp(metav1.NewMicroTime(tc.x), metav1.NewMicroTime(tc.y)),
+		"Microsecond (metav1.MicroTime <> metav1.MicroTime) comparison failed. Values:\n- x: %s\n- y: %s",
+		tc.x.String(), tc.y.String())
+	assert.Equal(t, tc.microsecond, tc.cmp(metav1.NewMicroTime(tc.x), tc.y),
+		"Microsecond (metav1.MicroTime <> stdlib) comparison failed. Values:\n- x: %s\n- y: %s",
+		tc.x.String(), tc.y.String())
+	assert.Equal(t, tc.microsecond, tc.cmp(tc.x, metav1.NewMicroTime(tc.y)),
+		"Microsecond (stdlib <> metav1.MicroTime) comparison failed. Values:\n- x: %s\n- y: %s",
+		tc.x.String(), tc.y.String())
+
+	assert.Equal(t, tc.second, tc.cmp(metav1.NewTime(tc.x), metav1.NewTime(tc.y)),
```

Comment:
can you also add a test for the case with `*metav1.Time`? does that work ok?

Reply:
(i have a feeling it will pass the compile check but panic)

Reply:
Yep, you are 100% correct. Cleaned up the logic + added pointer support (thanks @maiamcc) and added some (non-exhaustive) cases that test pointers of each type as well as pointer<>value.